{
  "cve_id": "CVE-2022-23583",
  "cve_desc": "Tensorflow is an Open Source Machine Learning Framework. A malicious user can cause a denial of service by altering a `SavedModel` such that any binary op would trigger `CHECK` failures. This occurs when the protobuf part corresponding to the tensor arguments is modified such that the `dtype` no longer matches the `dtype` expected by the op. In that case, calling the templated binary operator for the binary op would receive corrupted data, due to the type confusion involved. If `Tin` and `Tout` don't match the type of data in `out` and `input_*` tensors then `flat<*>` would interpret it wrongly. In most cases, this would be a silent failure, but we have noticed scenarios where this results in a `CHECK` crash, hence a denial of service. The fix will be included in TensorFlow 2.8.0. We will also cherrypick this commit on TensorFlow 2.7.1, TensorFlow 2.6.3, and TensorFlow 2.5.3, as these are also affected and still in supported range.",
  "repo": "tensorflow/tensorflow",
  "patch_hash": "a7c02f1a9bbc35473969618a09ee5f9f5d3e52d9",
  "patch_info": {
    "commit_hash": "a7c02f1a9bbc35473969618a09ee5f9f5d3e52d9",
    "repo": "tensorflow/tensorflow",
    "commit_url": "https://github.com/tensorflow/tensorflow/commit/a7c02f1a9bbc35473969618a09ee5f9f5d3e52d9",
    "files": [
      "tensorflow/core/kernels/cwise_ops_common.h"
    ],
    "message": "Validate real and expected type of arguments to cwise ops.\n\nWithout this validation, it is possible to trigger a `CHECK`-fail denial of service.\n\nThis is a rollforward of a previous commit which was rolled back as it was relying on RTTI. This time we don't use RTTI, we replace `typeid(Tin).name()` with a double function call, `DataTypeString(DataTypeToEnum<Tin>::v())`.\n\nPiperOrigin-RevId: 409340416\nChange-Id: I96080b2796729a3a9b65e7c68307ac276070f2f0",
    "before_after_code_files": [
      "tensorflow/core/kernels/cwise_ops_common.h||tensorflow/core/kernels/cwise_ops_common.h"
    ]
  },
  "patch_diff": {
    "tensorflow/core/kernels/cwise_ops_common.h||tensorflow/core/kernels/cwise_ops_common.h": [
      "File: tensorflow/core/kernels/cwise_ops_common.h -> tensorflow/core/kernels/cwise_ops_common.h",
      "--- Hunk 1 ---",
      "[Context before]",
      "88:   void Compute(OpKernelContext* ctx) override {",
      "89:     const Tensor& input_0 = ctx->input(0);",
      "90:     const Tensor& input_1 = ctx->input(1);",
      "91:     const Device& eigen_device = ctx->eigen_device<Device>();",
      "92:     bool error = false;",
      "93:     bool* const error_ptr = Functor::has_errors ? &error : nullptr;",
      "",
      "[Removed Lines]",
      "[None]",
      "",
      "[Added Lines]",
      "90:     OP_REQUIRES(ctx, input_0.dtype() == DataTypeToEnum<Tin>::v(),",
      "91:                 errors::InvalidArgument(",
      "92:                     \"Expected tensor of type \",",
      "93:                     DataTypeString(DataTypeToEnum<Tin>::v()), \" but got type \",",
      "94:                     DataTypeString(input_0.dtype())));",
      "96:     OP_REQUIRES(ctx, input_1.dtype() == DataTypeToEnum<Tin>::v(),",
      "97:                 errors::InvalidArgument(",
      "98:                     \"Expected tensor of type \",",
      "99:                     DataTypeString(DataTypeToEnum<Tin>::v()), \" but got type \",",
      "100:                     DataTypeString(input_1.dtype())));",
      "",
      "---------------"
    ]
  },
  "candidates": [
    {
      "candidate_hash": "29b71bf810c9d77078cc5d93191f7cee6421c65c",
      "candidate_info": {
        "commit_hash": "29b71bf810c9d77078cc5d93191f7cee6421c65c",
        "repo": "tensorflow/tensorflow",
        "commit_url": "https://github.com/tensorflow/tensorflow/commit/29b71bf810c9d77078cc5d93191f7cee6421c65c",
        "files": [
          "tensorflow/core/kernels/cwise_ops_common.h"
        ],
        "message": "Validate real and expected type of arguments to cwise ops.\n\nWithout this validation, it is possible to trigger a `CHECK`-fail denial of service.\n\nThis is a rollforward of a previous commit which was rolled back as it was relying on RTTI. This time we don't use RTTI, we replace `typeid(Tin).name()` with a double function call, `DataTypeString(DataTypeToEnum<Tin>::v())`.\n\nPiperOrigin-RevId: 409340416\nChange-Id: I96080b2796729a3a9b65e7c68307ac276070f2f0",
        "before_after_code_files": [
          "tensorflow/core/kernels/cwise_ops_common.h||tensorflow/core/kernels/cwise_ops_common.h"
        ]
      },
      "candidate_patch_features": {
        "candidate_earlier_than_patch": 0,
        "diff_branch_same_aad": 1,
        "olp_code_files": {
          "patch": [
            "tensorflow/core/kernels/cwise_ops_common.h||tensorflow/core/kernels/cwise_ops_common.h"
          ],
          "candidate": [
            "tensorflow/core/kernels/cwise_ops_common.h||tensorflow/core/kernels/cwise_ops_common.h"
          ]
        }
      },
      "candidate_diff": {
        "tensorflow/core/kernels/cwise_ops_common.h||tensorflow/core/kernels/cwise_ops_common.h": [
          "File: tensorflow/core/kernels/cwise_ops_common.h -> tensorflow/core/kernels/cwise_ops_common.h",
          "--- Hunk 1 ---",
          "[Context before]",
          "88:   void Compute(OpKernelContext* ctx) override {",
          "89:     const Tensor& input_0 = ctx->input(0);",
          "90:     const Tensor& input_1 = ctx->input(1);",
          "91:     const Device& eigen_device = ctx->eigen_device<Device>();",
          "92:     bool error = false;",
          "93:     bool* const error_ptr = Functor::has_errors ? &error : nullptr;",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "90:     OP_REQUIRES(ctx, input_0.dtype() == DataTypeToEnum<Tin>::v(),",
          "91:                 errors::InvalidArgument(",
          "92:                     \"Expected tensor of type \",",
          "93:                     DataTypeString(DataTypeToEnum<Tin>::v()), \" but got type \",",
          "94:                     DataTypeString(input_0.dtype())));",
          "96:     OP_REQUIRES(ctx, input_1.dtype() == DataTypeToEnum<Tin>::v(),",
          "97:                 errors::InvalidArgument(",
          "98:                     \"Expected tensor of type \",",
          "99:                     DataTypeString(DataTypeToEnum<Tin>::v()), \" but got type \",",
          "100:                     DataTypeString(input_1.dtype())));",
          "",
          "---------------"
        ]
      }
    },
    {
      "candidate_hash": "ae47a00c64be123549f60840c5ce5b7faa008040",
      "candidate_info": {
        "commit_hash": "ae47a00c64be123549f60840c5ce5b7faa008040",
        "repo": "tensorflow/tensorflow",
        "commit_url": "https://github.com/tensorflow/tensorflow/commit/ae47a00c64be123549f60840c5ce5b7faa008040",
        "files": [
          "tensorflow/core/kernels/cwise_ops_common.h"
        ],
        "message": "Validate real and expected type of arguments to cwise ops.\n\nWithout this validation, it is possible to trigger a `CHECK`-fail denial of service.\n\nThis is a rollforward of a previous commit which was rolled back as it was relying on RTTI. This time we don't use RTTI, we replace `typeid(Tin).name()` with a double function call, `DataTypeString(DataTypeToEnum<Tin>::v())`.\n\nPiperOrigin-RevId: 409340416\nChange-Id: I96080b2796729a3a9b65e7c68307ac276070f2f0",
        "before_after_code_files": [
          "tensorflow/core/kernels/cwise_ops_common.h||tensorflow/core/kernels/cwise_ops_common.h"
        ]
      },
      "candidate_patch_features": {
        "candidate_earlier_than_patch": 0,
        "diff_branch_same_aad": 1,
        "olp_code_files": {
          "patch": [
            "tensorflow/core/kernels/cwise_ops_common.h||tensorflow/core/kernels/cwise_ops_common.h"
          ],
          "candidate": [
            "tensorflow/core/kernels/cwise_ops_common.h||tensorflow/core/kernels/cwise_ops_common.h"
          ]
        }
      },
      "candidate_diff": {
        "tensorflow/core/kernels/cwise_ops_common.h||tensorflow/core/kernels/cwise_ops_common.h": [
          "File: tensorflow/core/kernels/cwise_ops_common.h -> tensorflow/core/kernels/cwise_ops_common.h",
          "--- Hunk 1 ---",
          "[Context before]",
          "88:   void Compute(OpKernelContext* ctx) override {",
          "89:     const Tensor& input_0 = ctx->input(0);",
          "90:     const Tensor& input_1 = ctx->input(1);",
          "91:     const Device& eigen_device = ctx->eigen_device<Device>();",
          "92:     bool error = false;",
          "93:     bool* const error_ptr = Functor::has_errors ? &error : nullptr;",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "90:     OP_REQUIRES(ctx, input_0.dtype() == DataTypeToEnum<Tin>::v(),",
          "91:                 errors::InvalidArgument(",
          "92:                     \"Expected tensor of type \",",
          "93:                     DataTypeString(DataTypeToEnum<Tin>::v()), \" but got type \",",
          "94:                     DataTypeString(input_0.dtype())));",
          "96:     OP_REQUIRES(ctx, input_1.dtype() == DataTypeToEnum<Tin>::v(),",
          "97:                 errors::InvalidArgument(",
          "98:                     \"Expected tensor of type \",",
          "99:                     DataTypeString(DataTypeToEnum<Tin>::v()), \" but got type \",",
          "100:                     DataTypeString(input_1.dtype())));",
          "",
          "---------------"
        ]
      }
    },
    {
      "candidate_hash": "0824de13b6068cf8bec2b4332141d32e21b2adb9",
      "candidate_info": {
        "commit_hash": "0824de13b6068cf8bec2b4332141d32e21b2adb9",
        "repo": "tensorflow/tensorflow",
        "commit_url": "https://github.com/tensorflow/tensorflow/commit/0824de13b6068cf8bec2b4332141d32e21b2adb9",
        "files": [
          "tensorflow/core/kernels/cwise_ops_common.h"
        ],
        "message": "Validate real and expected type of arguments to cwise ops.\n\nWithout this validation, it is possible to trigger a `CHECK`-fail denial of service.\n\nThis is a rollforward of a previous commit which was rolled back as it was relying on RTTI. This time we don't use RTTI, we replace `typeid(Tin).name()` with a double function call, `DataTypeString(DataTypeToEnum<Tin>::v())`.\n\nPiperOrigin-RevId: 409340416\nChange-Id: I96080b2796729a3a9b65e7c68307ac276070f2f0",
        "before_after_code_files": [
          "tensorflow/core/kernels/cwise_ops_common.h||tensorflow/core/kernels/cwise_ops_common.h"
        ]
      },
      "candidate_patch_features": {
        "candidate_earlier_than_patch": 0,
        "diff_branch_same_aad": 1,
        "olp_code_files": {
          "patch": [
            "tensorflow/core/kernels/cwise_ops_common.h||tensorflow/core/kernels/cwise_ops_common.h"
          ],
          "candidate": [
            "tensorflow/core/kernels/cwise_ops_common.h||tensorflow/core/kernels/cwise_ops_common.h"
          ]
        }
      },
      "candidate_diff": {
        "tensorflow/core/kernels/cwise_ops_common.h||tensorflow/core/kernels/cwise_ops_common.h": [
          "File: tensorflow/core/kernels/cwise_ops_common.h -> tensorflow/core/kernels/cwise_ops_common.h",
          "--- Hunk 1 ---",
          "[Context before]",
          "88:   void Compute(OpKernelContext* ctx) override {",
          "89:     const Tensor& input_0 = ctx->input(0);",
          "90:     const Tensor& input_1 = ctx->input(1);",
          "91:     const Device& eigen_device = ctx->eigen_device<Device>();",
          "92:     bool error = false;",
          "93:     bool* const error_ptr = Functor::has_errors ? &error : nullptr;",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "90:     OP_REQUIRES(ctx, input_0.dtype() == DataTypeToEnum<Tin>::v(),",
          "91:                 errors::InvalidArgument(",
          "92:                     \"Expected tensor of type \",",
          "93:                     DataTypeString(DataTypeToEnum<Tin>::v()), \" but got type \",",
          "94:                     DataTypeString(input_0.dtype())));",
          "96:     OP_REQUIRES(ctx, input_1.dtype() == DataTypeToEnum<Tin>::v(),",
          "97:                 errors::InvalidArgument(",
          "98:                     \"Expected tensor of type \",",
          "99:                     DataTypeString(DataTypeToEnum<Tin>::v()), \" but got type \",",
          "100:                     DataTypeString(input_1.dtype())));",
          "",
          "---------------"
        ]
      }
    }
  ]
}